### PR TITLE
Support BPF_PROG_{QUERY,ATTACH,DETACH}

### LINKED ISSUE #3652: Add BPF_PROG_QUERY support
On my computer, running runc within rr yield the following error message:
```
Assertion `t->regs().syscall_result_signed() == -syscall_state.expect_errno' failed to hold.
Expected EINVAL for 'bpf' but got result 0 (errno SUCCESS); unknown bpf(cmd=16)
```

`bpf(cmd=16)` is `BPF_PROG_QUERY`, which I confirmed with strace is indeed used by runc v1.1.7 with kernel 6.6.0 and user namespaces disabled.

### PR DESCRIPTION
**Pull request by multun**
Fixes #3652

### CODE REVIEW

File: src/test/bpf_query.c [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+      atomic_puts("Skipping test because it requires CAP_SYS_ADMIN");
+      atomic_puts("EXIT-SUCCESS");
+      return 0;
+    }
+    if (errno == ENOENT) {
+      atomic_puts("No cgroup vfs mounted on /sys/fs/cgroup, skipping");
+      atomic_puts("EXIT-SUCCESS");
+      return 0;
+    }
+    test_assert(!"failed to create cgroup");
```
Comment (rocallahan) [inline src/test/bpf_query.c:49]: So how does this test work when you run multiple copies simultaneously? Can we create a unique cgroup per test?

Comment (rocallahan) [reply src/test/bpf_query.c:49]: Or can we use socket or perf eBPF programs instead so it's simpler to set up the right environment?

Comment (rocallahan) [reply src/test/bpf_query.c:49]: Or are cgroups the only way to test `BPF_PROG_ATTACH`?

File: src/record_syscall.cc [hunk 2]
```diff
@@ -2194,6 +2194,46 @@ static Switchable prepare_bpf(RecordTask* t,
...
+        syscall_state.mem_ptr_parameter(link_ids_p, buf_size, OUT);
+
+      if ((link_attach_flags_p + 1).as_int() <= end_addr)
+        syscall_state.mem_ptr_parameter(link_attach_flags_p, buf_size, OUT);
+      break;
+    }
+    case BPF_PROG_DETACH:
+    case BPF_PROG_ATTACH: {
+      auto attr_size = t->regs().arg3();
+      syscall_state.reg_parameter(2, attr_size, IN_OUT);
```
Comment (rocallahan) [inline src/record_syscall.cc:2234]: Do these syscalls actually modify memory in user space?

File: src/record_syscall.cc [hunk 3]
```diff
@@ -2194,6 +2194,46 @@ static Switchable prepare_bpf(RecordTask* t,
...
+
+      auto prog_ids_p = REMOTE_PTR_FIELD(attrp, query.prog_ids);
+      auto prog_attach_flags_p = REMOTE_PTR_FIELD(attrp, query.prog_attach_flags);
+      auto link_ids_p = REMOTE_PTR_FIELD(attrp, query.link_ids);
+      auto link_attach_flags_p = REMOTE_PTR_FIELD(attrp, query.link_attach_flags);
+
+      // for each output array, only track changes if the field is
+      // within the bounds of the user provided structure
+      if ((prog_ids_p + 1).as_int() <= end_addr)
+        syscall_state.mem_ptr_parameter(prog_ids_p, buf_size, OUT);
```
Comment (rocallahan) [inline src/record_syscall.cc:2219]: {} around if bodies

File: src/record_syscall.cc [hunk 4]
```diff
@@ -2194,6 +2194,46 @@ static Switchable prepare_bpf(RecordTask* t,
...
+      auto attr_size = t->regs().arg3();
+      auto attrp = syscall_state.reg_parameter(2, attr_size, IN_OUT)
+        .cast<typename Arch::bpf_attr>();
+      auto end_addr = attrp.as_int() + attr_size;
+
+      // if the offset of the prog_cnt is out of the buffer,
+      // the syscall will fail and we can't track anything
+      auto prog_cnt_p = REMOTE_PTR_FIELD(attrp, query.prog_cnt);
+      if ((prog_cnt_p + 1).as_int() > end_addr)
+        break;
```
Comment (rocallahan) [inline src/record_syscall.cc:2207]: {} around if body

File: src/record_syscall.cc (hunk 3 above)
Comment (rocallahan) [reply src/record_syscall.cc:2219]: `OUT` is the default so let's just leave that out

**Review by rocallahan — CHANGES_REQUESTED** [review]

File: src/test/bpf_query.c (hunk 1 above)
Comment (multun) [reply src/test/bpf_query.c:49]: > So how does this test work when you run multiple copies simultaneously? Can we create a unique cgroup per test?

I'll create a cgroup per test

> Or can we use socket or perf eBPF programs instead so it's simpler to set up the right environment?

I believe socket / perf programs use another API. Here the full list of attach types:
https://www.kernel.org/doc/html/latest/bpf/libbpf/program_types.html

Most are cgroup stuff and kernel probes. I wish one of the types could be attached to an easier to create fd, but can't find any. I believe setting up a kprobe would be slightly easier, but would also have concurrency issues.

**Comment by multun** [comment]
I've just noticed that just running ```rr record runc --help; rr replay`` does not work on my laptop, with or without this PR. I'll until I can test on a processor where this works (I'm pretty sure this works fine on my workstation)

**Comment by multun** [comment]
I've made a couple of changes:
 - I've added a remote_buffer class to make the bound checks more readable
 - I've changed the test so it uses a network namespace and a BPF flow dissector

File: src/record_syscall.cc [hunk 5]
```diff
@@ -2196,42 +2197,46 @@ static Switchable prepare_bpf(RecordTask* t,
...
+      if (attr_buf.contains(link_attach_flags_p)) {
+        syscall_state.mem_ptr_parameter(link_attach_flags_p, buf_size);
+      }
       break;
     }
     case BPF_PROG_DETACH:
     case BPF_PROG_ATTACH: {
       auto attr_size = t->regs().arg3();
-      syscall_state.reg_parameter(2, attr_size, IN_OUT);
+      syscall_state.reg_parameter(2, attr_size, IN);
```
Comment (rocallahan) [inline src/record_syscall.cc:2239]: We don't need this at all. You can just ignore it.

File: src/record_syscall.cc [hunk 6]
```diff
@@ -2196,42 +2197,46 @@ static Switchable prepare_bpf(RecordTask* t,
     }
     case BPF_PROG_QUERY: {
       auto attr_size = t->regs().arg3();
-      auto attrp = syscall_state.reg_parameter(2, attr_size, IN_OUT)
-        .cast<typename Arch::bpf_attr>();
-      auto end_addr = attrp.as_int() + attr_size;
+      auto attr_begin = syscall_state.reg_parameter(2, attr_size, IN_OUT);
+      auto attr_buf = remote_buffer<void>(attr_begin, attr_size);
```
Comment (rocallahan) [inline src/record_syscall.cc:2201]: How about just using `MemoryRange` instead of `remote_buffer`?

**Review by rocallahan — CHANGES_REQUESTED** [review]

**Comment by khuey** [comment]
It seems like we should probably be doing the thing bpf does where it checks that the space (if any) between sizeof(bpf_attr) and the size argument to the syscall is 0. Otherwise if these structs get extended underneath us on newer kernels rr will silently diverge.

File: src/record_syscall.cc (hunk 6 above)
Comment (multun) [reply src/record_syscall.cc:2201]: I... looked for it and couldn't find it :clown_face: thanks!

**Comment by multun** [comment]
@khuey I've implemented your suggestion in [041eba7](https://github.com/rr-debugger/rr/pull/3653/commits/041eba751a1cdb34342e30a38d405e6422a57d3d)

File: src/kernel_abi.h [hunk 7]
```diff
@@ -1781,6 +1781,29 @@ struct BaseArch : public wordsize,
       aligned_u64 line_info;
       __u32 line_info_cnt;
     };
+
+    struct { /* anonymous struct used by BPF_PROG_QUERY command */
+      union {
+        __u32   target_fd;      /* target object to query or ... */
+        __u32   target_ifindex; /* target ifindex */
```
Comment (rocallahan) [inline src/kernel_abi.h:1788]: Get rid of the extra spaces between the type and the field name, here and below.

**Review by rocallahan — CHANGES_REQUESTED** [review]
Getting close... sorry for the iterations here.

I'd prefer it if you just squash all these commits together for review and merging.

**Comment by rocallahan** [comment]
Thanks!

**Comment by multun** [comment]
@khuey can you try this out? #3663

**Pull request merged**